### ISSUE TITLE
Support the Language Server Protocol

### ISSUE THREAD

**Issue by ghost** [jmacdonald/amp#37]
Have you considered supporting the [language server protocol](https://microsoft.github.io/language-server-protocol/) (LSP)?

Supporting this directly in the editor might be a good idea since it would already cover a lot of the use cases that would otherwise be relegated to a plugin API. Pretty much all of the other major editors (atom, emacs, code, (neo)vim, etc.) have decent support for the LSP at this point and most major languages have good server implementations too.

**Comment by jmacdonald** [comment]
This is genius. 👌 

I love the idea of an integrated, general solution that isn't language-specific. It looks like RLS [supports it](https://internals.rust-lang.org/t/introducing-rust-language-server-source-release/4209), and preliminary support is available for [Ruby](https://github.com/mtsmfm/language_server-ruby), too.

If there is broad enough language support for this, I think this is a great direction for Amp.

**Comment by zaynetro** [comment]
Language support is rather reasonable: http://langserver.org/ (There is a table "Language servers")
The page might be a bit outdated. Nonetheless, there are a plenty of languages there.

As an example [sbt recently released support for Language Server protocol](https://github.com/sbt/sbt/releases/tag/v1.1.0).

Yep, RLS is really well working.

**Comment by korken89** [comment]
Any update on LSP support? For me this is the only thing missing to use amp day to day :)

Keep up the good work ^^

**Comment by jmacdonald** [comment]
Thanks, @korken89!

The priority for `0.6` is moving away from termbox, which will get us 24-bit color support, drop the Python build requirement, and remove a C library dependency. I'm thinking the focus for `0.7` may be LSP support. I'll do a bit of roadmap planning and will add that to the README; thanks for the nudge. :slightly_smiling_face:

**Comment by korken89** [comment]
Thanks for the update, can't wait to see where this goes!

**Comment by TotalKrill** [comment]
Just a small nudge here, I am in the exact same boat as @korken89, where LSP being the main blocker for me. Would love to nuke my NeoVim config to use something that does not have 10x different plugins that breaks on libboost update or something similar .

Very nice editor you have created! Cant wait to see how 0.7 turns out!

**Comment by jmacdonald** [comment]
I'd love to know what, specifically, folks are after with LSP support. I've often wanted "jump to definition", as navigating/reading code is something I'm doing fairly often. Would appreciate other people's perspectives before I start planning the implementation for this. :slightly_smiling_face:

**Comment by mzarnitsa** [comment]
Agree. This is one of the most important features of a dev env. Especially when reading someone else’s code. That and contextual auto suggestion.

**Comment by TotalKrill** [comment]
I constantly use tools to get autosuggestions about accessible methods and members on a structure, as well as their signatures, saves me a trip to the documentation quite often, as well as typing the entire function name. 

This usually keeps me from using abbreviated method names which is common in environments without autocomplete. 

The other thing which I do not know is LSP is syntax errors and compilation warning/error indication directly in the editor. 

Ofcourse I also use jump to definition and to jump back to where I was after. In vim I have ctrl+I (jump in) gotot definition and ctrl+o (jump out) to go back

**Comment by jrop** [comment]
@jmacdonald Here's why I want it:
* Type ahead (huge)
* Show types of variables (where the type is inferred) (huge)
* Show declaration
* Rename symbol
* Go to definition

**Comment by TotalKrill** [comment]
Is there any status update on this?

**Comment by nyaascii** [comment]
This might be helpful: [autozimu/LanguageClient-neovim](https://github.com/autozimu/LanguageClient-neovim)
It's a Language Server Protocol client written in Rust, although intended for neovim 😄

**Comment by greenfork** [comment]
My 2 cents on how I use LSP in the order from most to least important:
1. Go to definition
2. Lookup documentation or get an error on the current line as a pop-up message under cursor
3. Rename symbol
4. Semantic token highlighting (stock highlighting isn't even used, generally useful to get highlighting on defined variables to avoid typos and it is a superior highlighting method in all respects too)
5. Formatting (mostly when it is interpreted language and formatting isn't as fast as Zig or Go version, I used it with Elixir and it's pleasingly fast)

**Comment by SadSock** [comment]
Is there any status update on this?

**Comment by jmacdonald** [comment]
👋🏼 yep, `0.8` is shaping up to include a bunch of improvements around open mode and adding file browser interop, and `0.9` is going to be focused around language server support. Will report back once that's underway. 😉

**Comment by gregorni** [comment]
I currently use [Helix](https://helix-editor.com/), which has a model that I find quite convincing: It basically has one huge builtin [table](https://docs.helix-editor.com/lang-support.html) of filetypes and associated language servers. At runtime, it checks if the LS binary it needs is in the $PATH and if so, loads it.

I find this approach has multiple benefits:
- As a user, I have one go to web page that says what LS I need to install for which language.
- I just need to install the LS, no further plugins or anything
- Once the LS is installed, everything _just works_

I haven't talked to any Helix devs, but I can imagine for them it also has benefits:
- Installing the LS is offloaded to the user, Helix just has to check if it's present.
- It's easy to abstract the LS implementation in the code: Define an implementation once, and have a list with all the LSes you want to support.
- Helix devs can choose exactly which LSes they want to support, which makes troubleshooting and bug fixing easier.

I've used Helix for quite a while now, and I really don't mind this model. You might think about using a similar implementation.

Also, Helix is also written in Rust, maybe you can check out their LSP implementation and copy parts.
